QDQ removal around Resize (mode=linear) causes wrong numeric values

### Describe the issue

The optimization in https://github.com/microsoft/onnxruntime/blob/88336ffa92e188e0866799dc559a9b6154980cf3/onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc#L58
removes Q-DQ nodes around Resize and thus makes the Resize compute in int8.
This might be fine for nearest interpolation mode, but in linear interpolation mode it changes the numeric output values.

The mismatch does not happen when using `sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_DISABLE_ALL`.

### To reproduce

 ```
#!/usr/bin/env python3
import onnx
import onnx.reference
from numpy import array, float32
import numpy

model = """\
<
   ir_version: 8,
   opset_import: ["" : 19]
>
main_graph (float[4] Concat_output_0, float[1,1,1,2] Abs_output_0) 
=> (float[unk__0,unk__1,unk__2,unk__3] DequantizeLinear_2_output_0) 
{
   Cast_2_output_0 = Constant <value: tensor = int8 {0}> ()
   Cast_3_output_0 = Constant <value: tensor = float {1.0}> ()
   QuantizeLinear_1_output_0 = QuantizeLinear (Abs_output_0, Cast_3_output_0, Cast_2_output_0)
   DequantizeLinear_1_output_0 = DequantizeLinear (QuantizeLinear_1_output_0, Cast_3_output_0, Cast_2_output_0)
   Resize_output_0 = Resize <mode: string = "linear"> (DequantizeLinear_1_output_0, , Concat_output_0)
   Cast_4_output_0 = Constant <value: tensor = int8 {0}> ()
   Cast_5_output_0 = Constant <value: tensor = float {1.0}> ()
   QuantizeLinear_2_output_0 = QuantizeLinear (Resize_output_0, Cast_5_output_0, Cast_4_output_0)
   DequantizeLinear_2_output_0 = DequantizeLinear (QuantizeLinear_2_output_0, Cast_5_output_0, Cast_4_output_0)
}

"""

m = onnx.parser.parse_model(model)

inputs = {'Concat_output_0': array([1., 1., 1., 2.], dtype=float32), 
          'Abs_output_0': array([[[[0.0 , 1.0]]]],dtype=float32)}

out = onnx.reference.ReferenceEvaluator(m).run(None, inputs)[0]
print("onnx.reference", out)

import onnxruntime
sess_options = onnxruntime.SessionOptions()
#sess_options.log_severity_level = 0
# Uncommenting the next line fixes the numeric mismatches
# sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_DISABLE_ALL
sess_options.optimized_model_filepath = "opt.onnx"
onnxref = onnxruntime.InferenceSession(m.SerializeToString(), sess_options).run(None, inputs)[0]

print("onnxruntime", onnxref)

numpy.testing.assert_allclose(out, onnxref, rtol=1e-3, atol=1e-3)
```
shows a numeric mismatch:
```
onnx.reference [[[[0. 0. 1. 1.]]]]
onnxruntime [[[[0. 0. 0. 1.]]]]
AssertionError: 
Not equal to tolerance rtol=0.001, atol=0.001
```
If you look at `opt.onnx`, you see that the model has been transformed into 
![image](https://github.com/microsoft/onnxruntime/assets/93204396/313cac63-7eb6-4a4e-a3ed-bd83c15f74b0)
which is invalid because the linear interpolation cannot be done correctly in int8.

### Urgency

_No response_

### Platform

Linux

### OS Version

Ubuntu 22.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.18.1

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU

### Execution Provider Library Version

_No response_

--- a/onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selectors.cc
+++ b/onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selectors.cc
@@ -181,6 +181,17 @@ bool DropQDQNodeGroupSelector::Check(const GraphViewer& graph_viewer, const Node
     return false;
   }
 
+  // Resize with mode != "nearest" (linear/cubic) is not order-preserving on integers,
+  // so dropping the surrounding Q/DQ pair would change numerical results. Only the
+  // default nearest-neighbor mode is safe to fold. See issue #21319.
+  if (node.OpType() == "Resize") {
+    const auto* mode_attr = graph_utils::GetNodeAttribute(node, "mode");
+    const std::string mode = (mode_attr != nullptr) ? mode_attr->s() : "nearest";
+    if (mode != "nearest") {
+      return false;
+    }
+  }
+
   const Node& dq_node = *dq_nodes.front();
   const Node& q_node = *q_nodes.front();
 

--- a/onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selectors.cc
+++ b/onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selectors.cc
@@ -186,8 +186,7 @@ bool DropQDQNodeGroupSelector::Check(const GraphViewer& graph_viewer, const Node
   // default nearest-neighbor mode is safe to fold. See issue #21319.
   if (node.OpType() == "Resize") {
     const auto* mode_attr = graph_utils::GetNodeAttribute(node, "mode");
-    const std::string mode = (mode_attr != nullptr) ? mode_attr->s() : "nearest";
-    if (mode != "nearest") {
+    if (mode_attr != nullptr && mode_attr->s() != "nearest") {
       return false;
     }
   }

--- a/onnxruntime/test/optimizer/qdq_transformer_test.cc
+++ b/onnxruntime/test/optimizer/qdq_transformer_test.cc
@@ -2034,6 +2034,44 @@ TEST(QDQTransformerTests, Resize_No_Fusion) {
   test_case({1, 8, 64, 64}, {4}, {1, 4, 128, 128}, 1, true /*use_contrib_qdq*/);
 }
 
+TEST(QDQTransformerTests, Resize_NonNearest_No_Fusion) {
+  // Resize with mode="linear" or mode="cubic" is not order-preserving on integers.
+  // The surrounding Q/DQ pair must NOT be dropped. See issue #21319.
+  auto test_case = [&](const std::vector<int64_t>& input1_shape,
+                       const std::vector<int64_t>& sizes_shape,
+                       const std::string& mode,
+                       bool use_contrib_qdq = false) {
+    auto check_graph = [&](InferenceSessionWrapper& session) {
+      auto op_to_count = CountOpsInGraph(session.GetGraph());
+      const QDQOpKeys qdq_keys = GetQDQOpKeys(use_contrib_qdq);
+      EXPECT_EQ(op_to_count["Resize"], 1);
+      EXPECT_EQ(op_to_count[qdq_keys.quantize_linear], 1);
+      EXPECT_EQ(op_to_count[qdq_keys.dequantize_linear], 1);
+    };
+
+    TransformerTester(BuildQDQResizeTestCase(input1_shape,
+                                             sizes_shape,
+                                             mode,                  // mode (linear or cubic)
+                                             "half_pixel",          // coordinate_transformation_mode
+                                             "round_prefer_floor",  // nearest_mode (unused for linear/cubic)
+                                             false,                 // add_dq_output_float
+                                             use_contrib_qdq),
+                      check_graph,
+                      TransformerLevel::Level1,
+                      TransformerLevel::Level2);
+  };
+
+  RandomValueGenerator rand_gen{optional<RandomValueGenerator::RandomSeedType>{2345}};
+  // mode="linear"
+  test_case({2, 13, 12, 37}, rand_gen.Uniform<int64_t>(std::vector<int64_t>{4}, 1, 16), "linear");
+  test_case({2, 13, 12, 37}, rand_gen.Uniform<int64_t>(std::vector<int64_t>{4}, 1, 16), "linear",
+            true /*use_contrib_qdq*/);
+  // mode="cubic"
+  test_case({2, 13, 12, 37}, rand_gen.Uniform<int64_t>(std::vector<int64_t>{4}, 1, 16), "cubic");
+  test_case({2, 13, 12, 37}, rand_gen.Uniform<int64_t>(std::vector<int64_t>{4}, 1, 16), "cubic",
+            true /*use_contrib_qdq*/);
+}
+
 TEST(QDQTransformerTests, ResizeReshapeSqueezeUnsqueeze) {
   auto test_case = [&](const std::vector<int64_t>& input_shape,
                        const std::vector<int64_t>& sizes_shape,

--- a/onnxruntime/test/optimizer/qdq_transformer_test.cc
+++ b/onnxruntime/test/optimizer/qdq_transformer_test.cc
@@ -2061,15 +2061,58 @@ TEST(QDQTransformerTests, Resize_NonNearest_No_Fusion) {
                       TransformerLevel::Level2);
   };
 
-  RandomValueGenerator rand_gen{optional<RandomValueGenerator::RandomSeedType>{2345}};
+  // Use fixed output sizes where batch and channel are unchanged (linear/cubic require scale=1 on batch/channel).
   // mode="linear"
-  test_case({2, 13, 12, 37}, rand_gen.Uniform<int64_t>(std::vector<int64_t>{4}, 1, 16), "linear");
-  test_case({2, 13, 12, 37}, rand_gen.Uniform<int64_t>(std::vector<int64_t>{4}, 1, 16), "linear",
-            true /*use_contrib_qdq*/);
+  test_case({1, 8, 64, 64}, {1, 8, 128, 128}, "linear");
+  test_case({1, 8, 64, 64}, {1, 8, 128, 128}, "linear", true /*use_contrib_qdq*/);
   // mode="cubic"
-  test_case({2, 13, 12, 37}, rand_gen.Uniform<int64_t>(std::vector<int64_t>{4}, 1, 16), "cubic");
-  test_case({2, 13, 12, 37}, rand_gen.Uniform<int64_t>(std::vector<int64_t>{4}, 1, 16), "cubic",
-            true /*use_contrib_qdq*/);
+  test_case({1, 8, 64, 64}, {1, 8, 128, 128}, "cubic");
+  test_case({1, 8, 64, 64}, {1, 8, 128, 128}, "cubic", true /*use_contrib_qdq*/);
+}
+
+TEST(QDQTransformerTests, Resize_DefaultMode_Fusion) {
+  // A Resize node with no "mode" attribute defaults to "nearest" per the ONNX spec.
+  // The selector must treat a missing mode attribute the same as mode="nearest" and
+  // allow Q/DQ to be dropped. See issue #21319.
+  auto test_case = [&](const std::vector<int64_t>& input1_shape,
+                       const std::vector<int64_t>& sizes_data,
+                       bool use_contrib_qdq = false) {
+    auto build_test_case = [&](ModelTestBuilder& builder) {
+      auto* input1_arg = builder.MakeInput<uint8_t>(input1_shape,
+                                                    std::numeric_limits<uint8_t>::min(),
+                                                    std::numeric_limits<uint8_t>::max());
+      auto* roi = builder.MakeInitializer<float>({0}, {});
+      auto* scales = builder.MakeInitializer<float>({0}, {});
+      auto* sizes = builder.Make1DInitializer<int64_t>(sizes_data);
+      auto* output_arg = builder.MakeOutput();
+
+      // add DQ
+      auto* dq_output = builder.MakeIntermediate();
+      builder.AddDequantizeLinearNode<uint8_t>(input1_arg, .003f, 1, dq_output, use_contrib_qdq);
+
+      // add Resize with no mode attribute (ONNX default = "nearest")
+      auto* resize_output = builder.MakeIntermediate();
+      builder.AddNode("Resize", {dq_output, roi, scales, sizes}, {resize_output});
+
+      // add Q
+      builder.AddQuantizeLinearNode<uint8_t>(resize_output, .003f, 1, output_arg, use_contrib_qdq);
+    };
+
+    auto check_graph = [&](InferenceSessionWrapper& session) {
+      auto op_to_count = CountOpsInGraph(session.GetGraph());
+      const QDQOpKeys qdq_keys = GetQDQOpKeys(use_contrib_qdq);
+      EXPECT_EQ(op_to_count["Resize"], 1);
+      EXPECT_EQ(op_to_count[qdq_keys.quantize_linear], 0);
+      EXPECT_EQ(op_to_count[qdq_keys.dequantize_linear], 0);
+    };
+
+    TransformerTester(build_test_case, check_graph,
+                      TransformerLevel::Level1,
+                      TransformerLevel::Level2);
+  };
+
+  test_case({1, 8, 64, 64}, {1, 8, 128, 128});
+  test_case({1, 8, 64, 64}, {1, 8, 128, 128}, true /*use_contrib_qdq*/);
 }
 
 TEST(QDQTransformerTests, ResizeReshapeSqueezeUnsqueeze) {

